### DISCUSSION TITLE
Matmul changes order of loops

### DISCUSSION THREAD

**Discussion by gautam-e** [modular/modular#351] [Q&A] [answered]
Is there a reason why the order of the loops are different? 
The docs claim to copy-paste the code but have a look at the inner loops:  

```
# This exactly the same Python implementation, 
# but is infact Mojo code!
def matmul_untyped(C, A, B):
    for m in range(C.rows):
        for n in range(C.cols): #<-- Here1
            for k in range(A.cols): #<-- Here2
                C[m, n] += A[m, k] * B[k, n]
```

```
# Note that C, A, and B have types.
fn matmul_naive(C: Matrix, A: Matrix, B: Matrix):
    for m in range(C.rows):
        for k in range(A.cols): #<-- Here2
            for n in range(C.cols): #<-- Here1
                C[m, n] += A[m, k] * B[k, n]
```

**Comment by lattner** [comment]
@abduld ?

**Comment by abduld** [comment, marked as answer]
The order 

```
    for m in range(C.rows):
        for k in range(A.cols):
            for n in range(C.cols):
                C[m, n] += A[m, k] * B[k, n]
```

is more efficient and was changed in https://docs.modular.com/mojo/changelog.html#changed-2 , but it looks like we missed a few cases. Thanks for the sharp eye. Can you please file an issue so that this will be updated in the next rollout

**Reply by gautam-e** [reply]
Sure. Will do it by the end of the day (Central Europe). 

BTW, I see that this change prepares for the vectorisation of the innermost loop, but I am not sure if it's really a `dot` product that's happening as the vectorised version suggests. Let me think that through though...
